### ISSUE TITLE
 Exception calling "GetHierarchy" with "3" argument(s)

### ISSUE THREAD

**Issue by BoltCitizen** [theohbrothers/ConvertOneNote2MarkDown#101]
## Bug

Script finishes with error message.

## Discussion

```
VERBOSE: The object written to the pipeline is an instance of the type
"Microsoft.Office.Interop.OneNote.Application2Class" from the component's primary interoperability assembly. If this
type exposes different members than the IDispatch members, scripts that are written to work with this object might not
work if the primary interoperability assembly is not installed.
Convert-OneNote2MarkDown : Exception calling "GetHierarchy" with "3" argument(s): "Unable to cast COM object of type
'Microsoft.Office.Interop.OneNote.Application2Class' to interface type
'Microsoft.Office.Interop.OneNote.IApplicationCOM'. This operation failed because the QueryInterface call on the COM
component for the interface with IID '{452AC71A-B655-4967-A208-A4CC39DD7949}' failed due to the following error:
Interface not registered (Exception from HRESULT: 0x80040155)."
At C:\OneNote2MarkDown\ConvertOneNote2MarkDown-master\ConvertOneNote2MarkDown-v2.ps1:1373 char:5
+     Convert-OneNote2MarkDown @params
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Convert-OneNote2MarkDown

Convert-OneNote2MarkDown : at Get-OneNoteHierarchy,
C:\OneNote2MarkDown\ConvertOneNote2MarkDown-master\ConvertOneNote2MarkDown-v2.ps1: line 548
at Convert-OneNote2MarkDown, C:\OneNote2MarkDown\ConvertOneNote2MarkDown-master\ConvertOneNote2MarkDown-v2.ps1: line
1312
at <ScriptBlock>, C:\OneNote2MarkDown\ConvertOneNote2MarkDown-master\ConvertOneNote2MarkDown-v2.ps1: line 1373
at <ScriptBlock>, <No file>: line 1
At C:\OneNote2MarkDown\ConvertOneNote2MarkDown-master\ConvertOneNote2MarkDown-v2.ps1:1373 char:5
+     Convert-OneNote2MarkDown @params
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Convert-OneNote2MarkDown
```

## Environment

Windows Surface Pro 6
Windows 10
Office 2019 Professional Plus
Word, OneNote and Powershell open in admin mode
Security Scope Bypass already done

**Comment by leojonathanoh** [comment]
@BoltCitizen thanks for the bug report! 

This appears to be a bug specific to some API change in OneNote 2019, [in the second argument of `onenote.GetHierarchy()` method](https://github.com/theohbrothers/ConvertOneNote2MarkDown/blob/840855b1973031cc4cc1b2b344a9a000898147b5/ConvertOneNote2MarkDown-v2.ps1#L548). I haven't actually tried 2019 myself, I'll try to find out the cause in the developer docs.

**Comment by leojonathanoh** [comment]
Reading up on the `Office 2019 Professional Plus` https://www.microsoft.com/en-us/microsoft-365/p/office-professional-2019/CFQ7TTC0K7C5/?activetab=pivot%3aoverviewtab, it doesn't actually include Onenote desktop, so it must have been installed separately. Could you check what version of OneNote desktop version you are running?

**Comment by leojonathanoh** [comment]
related: https://github.com/aviaryan/VSCodeNotebook/issues/14, at that time `v1.0` was just released. So this error affected certain users since [v1.0](https://github.com/theohbrothers/ConvertOneNote2MarkDown/blob/v1.0/ConvertOneNote2MarkDown.ps1#L21) in the `GetHierarchy()` method.

**Comment by leojonathanoh** [comment]
@BoltCitizen  on my Windows 10 1909 machine where it usually works for me, is the very latest Onenote desktop 2016 as of today, downloaded from https://www.onenote.com/Download:

![20210826T044240-1629924160](https://user-images.githubusercontent.com/11932107/130862486-5ae9e6b2-9039-431c-8363-65b3570e0c05.png)

I spun up a brand new test Windows 10 1809 VM and installed OneNote from the same link, and it also installed the correct version, and could run the script with no errors about `GetHierarchy()`:

![20210826T044145-1629924105](https://user-images.githubusercontent.com/11932107/130863114-2f77e450-2185-4924-b369-4eac2db96067.png)

So perhaps you can try installing OneNote 2016 desktop from that URL. Let me know if that works for you.

**Comment by BoltCitizen** [comment]
tl;dr
Working on it.

I started to love all things Open Source and hating Microsoft. So I bought a new laptop with Linux and switching now, which includes OneNote to Zettlr.  You are right, my OneNote is outdated. I cant install the new version, I cannot uninstall the old version either. Turns out my entire Windows is outdated and due to a bug I cannot update it. Now I need to reinstall Windows completely, just to update OneNote, to finally get rid of Microsoft once and for all. 

Thank you so much for your support.

**Comment by leojonathanoh** [comment]
@BoltCitizen we do seem to be on a similar path, moving away from certain systems. Hope that the new installation sorts out things for you 😃

**Comment by BoltCitizen** [comment]
- Reinstalled und updated Windows 10.
- Uninstalled Office and reinstalled only OneNote following your link to the latest version: 
![image](https://user-images.githubusercontent.com/65156740/131230041-efcac29a-75b6-492c-b653-99a1390b4f63.png)
- Ran the script and got a different error, but this time its about **4** arguments, instead of 3: 
```
- Converting notebook 'Test'... (Ignoring deleted notes)

Building conversion configuration for Test [Notebook]
# Building conversion configuration for Test Section [Section]
## Building conversion configuration for Test Page [Page]
## Test Page [Page]
Failed to convert page: Test-Section\Test-Page. Reason: Error while publishing page to docx file Test Page: Exception calling "Publish" with "4" argument(s): "Class not registered

Class not registered
"
Convert-OneNotePage : Failed to convert page: Test-Section\Test-Page. Reason: Error while publishing page to docx file Test Page: Exception
calling "Publish" with "4" argument(s): "Class not registered
Class not registered
"
At C:\OneNote2MarkDown\ConvertOneNote2MarkDown-master\ConvertOneNote2MarkDown-v2.ps1:1337 char:257
+ ... onConfigs | Convert-OneNotePage -OneNoteConnection $OneNote -Config $ ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Convert-OneNotePage


Done converting notebook 'Test' with 1 notes.
Cleaning up...
Conversion errors:
Failed to convert page: Test-Section\Test-Page. Reason: Error while publishing page to docx file Test Page: Exception calling "Publish" with "4" argument(s): "Class not registered

Class not registered
"
Exiting.
```

**Comment by leojonathanoh** [comment]
@joeltimothyoh do you have MS office installed? I think its related to Word not being installed.

**Comment by leojonathanoh** [comment]
do you have MS office installed? I think its related to Word not being installed. @BoltCitizen 

much like the VM i spun up to test OneNote which also didn't have Word installed, see screenshot in https://github.com/theohbrothers/ConvertOneNote2MarkDown/issues/101#issuecomment-905867512

**Comment by leojonathanoh** [comment]
if this installing Word works for you, i'll probably have to update the readme prerequisites with an explicit mention of requiring Word

**Comment by leojonathanoh** [comment]
i believe that if you click File > export > page and don't see `.docx` under one of the available formats, then the script would also not be able to publish the onenote pages as `.docx` (hence the error).

**Comment by BoltCitizen** [comment]
I can export pages to .docx just fine with OneNote, without having Word installed.

Are you sure Word is required?

**Comment by leojonathanoh** [comment]
@BoltCitizen interesting. Let me try in my VM.

**Comment by BoltCitizen** [comment]
FYI
After reinstalling Windows 10, OneNote was still in the old state and there was no way to update it. It was also not listed as a separate program under add/remove programs. After uninstalling Office 2019 it was gone. It WAS part of the office package.

**Comment by leojonathanoh** [comment]
> FYI
> After reinstalling Windows 10, OneNote was still in the old state and there was no way to update it. It was also not listed as a separate program under add/remove programs. After uninstalling Office 2019 it was gone. It WAS part of the office package.

yeah that was probably a "locked-in" version of onenote bound to office 2019, although from the link in https://github.com/theohbrothers/ConvertOneNote2MarkDown/issues/101#issuecomment-905830140, onenote is not part of Office 2019. But overall i believe it should not matter, so long as you install the standalone version of onenote 2016 from the link in https://github.com/theohbrothers/ConvertOneNote2MarkDown/issues/101#issuecomment-905867512

@BoltCitizen in my brand new test VM I don't see an option to export the page as `.docx`, without Word installed. 

<img width="981" alt="image" src="https://user-images.githubusercontent.com/11932107/131244265-580799c6-d028-4061-968c-f80965714f30.png">

**Comment by leojonathanoh** [comment]
id think you might have to reinstall office 2019, i don't have a trial Office 365 to test it in my VM so i can't test whether installing Word alone will be enough.

**Comment by BoltCitizen** [comment]
TL;DR
It seams to work now, script is running.

- With only OneNote installed as per your link, it is in the latest version 20250, no Word installed, I can indeed export manually to .docx files, but running the script I get the error as shown above: `"Publish" with "4" argument(s): "Class not registered`
- When installing office 365, including Word, OneNote is automatically downgraded to version 20238, I do not know why. But, it seams to work now. Script is runnning. Hoping for the best.

Any recommendations for best conversion settings for Zettlr?

**Comment by leojonathanoh** [comment]
> With only OneNote installed as per your link, it is in the latest version 20250, no Word installed, I can indeed export manually to .docx files, but running the script I get the error as shown above: "Publish" with "4" argument(s): "Class not registered

nice to know. Somethings different, perhaps there's some lingering version of office even after your reinstallation of Windows. But if you said you uninstalled office 2019, but still see an option to export onenote pages as `.docx`, I can't really think of why though 😕 

 > When installing office 365, including Word, OneNote is automatically downgraded to version 20238, I do not know why. But, it seams to work now. Script is runnning. Hoping for the best.

yeah it does seem like the installer for 365 might install a different version. I think the main version is under File > Account > About Onenote https://github.com/theohbrothers/ConvertOneNote2MarkDown/issues/102#issuecomment-906127501, which might show a slightly different version e.g. `16.0.14228.20216` when the `onenote.exe`'s version is `16.0.14228.20250`, but overall it doesn't really matter so long as it's Onenote 2016 :)

You can do a dry run first, if it passes then generally it should be fine and have no errors.

> Any recommendations for best conversion settings for Zettlr?

Never heard of it 👀

**Comment by BoltCitizen** [comment]
Well then I recommend you to take a look at Zettlr :)
FOSS, everything in mark-down, individual files with folder structure, beautifull to look at

**Comment by leojonathanoh** [comment]
thinking about it now, on the surface onenote appears standalone tool when it is used by itself, but only shows it's not when migrating away from it - seeing now the requirement of needing a trial version of 365 (just for Word for `.docx`) which requires the setup of payment methods (i.e. credit card etc) - and is really a way of locking one in the ecosystem.

**Comment by leojonathanoh** [comment]
> Well then I recommend you to take a look at Zettlr :)

I'm stil migrating my notebooks. Will check it out :)

> FOSS, everything in mark-down, individual files with folder structure, beautifull to look at

Love markdown too. Best thing is can be consumed on the web.

**Comment by nikopen** [comment]
@leojonathanoh having a similar issue!

```
Convert-OneNote2MarkDown : Exception calling "GetHierarchy" with "3" argument(s): "Unable to cast COM object of type
'Microsoft.Office.Interop.OneNote.Application2Class' to interface type
'Microsoft.Office.Interop.OneNote.IApplicationCOM'. This operation failed because the QueryInterface call on the COM
component for the interface with IID '{452AC71A-B655-4967-A208-A4CC39DD7949}' failed due to the following error:
Interface not registered (Exception from HRESULT: 0x80040155)."
At D:\ConvertOneNote2MarkDown\ConvertOneNote2MarkDown-v2.ps1:1423 char:5
+     Convert-OneNote2MarkDown @params
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Convert-OneNote2MarkDown

Convert-OneNote2MarkDown : at Get-OneNoteHierarchy, D:\ConvertOneNote2MarkDown\ConvertOneNote2MarkDown-v2.ps1: line 568
at Convert-OneNote2MarkDown, D:\ConvertOneNote2MarkDown\ConvertOneNote2MarkDown-v2.ps1: line 1362
at <ScriptBlock>, D:\ConvertOneNote2MarkDown\ConvertOneNote2MarkDown-v2.ps1: line 1423
at <ScriptBlock>, <No file>: line 1
At D:\ConvertOneNote2MarkDown\ConvertOneNote2MarkDown-v2.ps1:1423 char:5
+     Convert-OneNote2MarkDown @params
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
    + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,Convert-OneNote2MarkDown
    ```
    
onenote 2016 latest, word 2019, powershell 5.1, admin everything.

**Comment by leojonathanoh** [comment]
@nikopen hmm that's strange. Could you ensure that onenote and word are the desktop versions of office, post the output of `$PSVersionTable`, `[System.Environment]::OSVersion.Version`, and post the script log with `-Verbose`

**Comment by nikopen** [comment]
@leojonathanoh yes they are! word 2019 proplus, onenote 2016 desktop latest, Win 10 21382

```
Name                           Value
----                           -----
PSVersion                      5.1.21382.1
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.21382.1
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
```

```
VERBOSE: The object written to the pipeline is an instance of the type
"Microsoft.Office.Interop.OneNote.Application2Class" from the component's primary interoperability assembly. If this
type exposes different members than the IDispatch members, scripts that are written to work with this object might not
work if the primary interoperability assembly is not installed.
```

Any way to just bypass all these?  
Can the script run on the .one files directly?  
Or better, any reliable way to just spin up a VM and run it there ? if yes it should be scriptable

**Comment by leojonathanoh** [comment]
@nikopen hmm I believe that your case seems similar to OP, where the bundled OneNote 2016 with the Office 2019 somehow doesn't work properly. You might have to uninstall Office 2019, which will remove OneNote. Then reinstall OneNote 2016 following https://github.com/theohbrothers/ConvertOneNote2MarkDown/issues/101#issuecomment-905867512, then install Word 2016 and above. 

> Any way to just bypass all these?

This is just a warning, should not matter.

>  Can the script run on the .one files directly?

No, the script uses the OneNote Object Model APIl

> Or better, any reliable way to just spin up a VM and run it there ? if yes it should be scriptable

Yes, the VM just has to have the prerequisites installed.

Overall I'd suggest you spin up a VM, install OneNote 2016 there, with Word 2016, and do the conversion there instead of the host system, so you keep your Surface machine intact.

**Comment by leojonathanoh** [comment]
See the process here: https://github.com/theohbrothers/ConvertOneNote2MarkDown/issues/101#issuecomment-907755120

**Comment by stinos** [comment]
Just a note for people running into this exception: another cause is mixing 32bit/64bit processes. So if you have the 32bit version of Office, make sure to launch a 32bit Powershell version (e.g. Windows Powershell ISE (x86)').
